### PR TITLE
[core] [C++] Simplify syncer unit test

### PR DESCRIPTION
**Pull request by dentiny**
See inline comments for details.

### CODE REVIEW

File: src/ray/common/test/ray_syncer_test.cc [hunk 1]
```diff
@@ -306,14 +306,10 @@ struct SyncerServerTest {
...
 
     while (duration_cast<seconds>(steady_clock::now() - start).count() <= time_s) {
-      std::promise<bool> p;
-      auto f = p.get_future();
-      io_context.post([&p, predicate]() mutable { p.set_value(predicate()); }, "TEST");
-      if (f.get()) {
+      if (predicate()) {
         return true;
-      } else {
-        std::this_thread::sleep_for(1s);
```
Comment (dentiny) [inline src/ray/common/test/ray_syncer_test.cc:315]: local grpc goes through loopback device, at worst is no worse than a fs access, which is 10s of microseconds magnitude; sleep for a second is a waste of testing time.

File: src/ray/common/test/ray_syncer_test.cc [hunk 2]
```diff
@@ -306,14 +306,10 @@ struct SyncerServerTest {
     auto start = steady_clock::now();
 
     while (duration_cast<seconds>(steady_clock::now() - start).count() <= time_s) {
-      std::promise<bool> p;
-      auto f = p.get_future();
-      io_context.post([&p, predicate]() mutable { p.set_value(predicate()); }, "TEST");
-      if (f.get()) {
```
Comment (dentiny) [inline src/ray/common/test/ray_syncer_test.cc:312]: `std::future<T>::get` is a blocking call, I don't see any meaning to use future / promise / io_context here.

Comment (rynewang) [reply src/ray/common/test/ray_syncer_test.cc:312]: I think the point is that it wants the predicate() func to be run in the ioctx thread, while also wants the post to be sync (blocking). If you look at callers of WaitUntil, many of them access ray_syncer_ which may not be thread safe.

Comment (dentiny) [reply src/ray/common/test/ray_syncer_test.cc:312]: OK I see, we need to queue the functor to make it executed sequentially.

**Comment by rynewang** [comment]
let's not focus on these tidying too much...

**Comment by dentiny** [comment]
> let's not focus on these tidying too much...

Yeah I spent several hours on the caltrain, would like to do some small tidy work instead of doing nothing...

**Pull request closed without merging**